### PR TITLE
Override the key to avoid CookieTheftException

### PR DESCRIPTION
**Pull request by XieEDeHeiShou**
People may confused why those `token.keyhash()` not equals, for they forget to config the key twice.
So, override the key for them, make it more convenient to custom RememberMeServices.
See #4140;

### CODE REVIEW

**Comment by pivotal-issuemaster** [comment]
@XieEDeHeiShou Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-projects/spring-security&pullRequestId=5509)!

[Click here](https://cla.pivotal.io/sync/spring?repositoryId=spring-projects/spring-security&pullRequestId=5509) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

**Comment by pivotal-issuemaster** [comment]
@XieEDeHeiShou Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-projects/spring-security&pullRequestId=5509)!

**Comment by XieEDeHeiShou** [comment]
The CI says it cannot get a jar from spring repository and then stopped with error:

```
* What went wrong:
Could not resolve all files for configuration ':spring-security-messaging:optional'.
> Could not download spring-websocket.jar (org.springframework:spring-websocket:5.1.0.BUILD-SNAPSHOT:20180713.230306-259)
   > Could not get resource 'https://repo.spring.io/libs-snapshot/org/springframework/spring-websocket/5.1.0.BUILD-SNAPSHOT/spring-websocket-5.1.0.BUILD-20180713.230306-259.jar'.
      > Connection reset
```

I checked the file manually, exists. 
Hope this pr could pass.

File: config/src/main/java/org/springframework/security/config/annotation/web/configurers/RememberMeConfigurer.java [hunk 1]
```diff
@@ -267,6 +267,19 @@ public void init(H http) throws Exception {
 		validateInput();
 		String key = getKey();
 		RememberMeServices rememberMeServices = getRememberMeServices(http, key);
+		if (rememberMeServices instanceof AbstractRememberMeServices) {
+			 /*
```
Comment (eleftherias) [inline config/src/main/java/org/springframework/security/config/annotation/web/configurers/RememberMeConfigurer.java]: Please remove the comments. The code should be clear enough that comments are not needed.

File: config/src/main/java/org/springframework/security/config/annotation/web/configurers/RememberMeConfigurer.java [hunk 2]
```diff
@@ -267,6 +267,19 @@ public void init(H http) throws Exception {
...
+         	  * To avoid CookieTheftException in PersistentTokenBasedRememberMeServices.processAutoLoginCookie(...),
+         	  * override the key or config it twice like below:
+         	  *
+         	  * <code>
+         	  *     http.rememberMe()
+         	  *      .key(key)
+         	  *      .rememberMeServices(new PersistentTokenBasedRememberMeServices(key, userDetailsService, tokenRepository));
+         	  * </code>
+         	  */
+			key = ((AbstractRememberMeServices) rememberMeServices).getKey();
```
Comment (eleftherias) [inline config/src/main/java/org/springframework/security/config/annotation/web/configurers/RememberMeConfigurer.java]: We do not want to override the key if the user has specified it in `.key("...")`. We should check if the key is null and only then use the one from `AbstractRememberMeServices`.

File: config/src/main/java/org/springframework/security/config/annotation/web/configurers/RememberMeConfigurer.java (hunk 1 above)
Comment (XieEDeHeiShou) [reply config/src/main/java/org/springframework/security/config/annotation/web/configurers/RememberMeConfigurer.java]: Removed

File: config/src/main/java/org/springframework/security/config/annotation/web/configurers/RememberMeConfigurer.java (hunk 2 above)
Comment (XieEDeHeiShou) [reply config/src/main/java/org/springframework/security/config/annotation/web/configurers/RememberMeConfigurer.java]: Done

File: config/src/main/java/org/springframework/security/config/annotation/web/configurers/RememberMeConfigurer.java [hunk 3]
```diff
@@ -267,6 +267,11 @@ public void init(H http) throws Exception {
 		validateInput();
 		String key = getKey();
 		RememberMeServices rememberMeServices = getRememberMeServices(http, key);
+		if (key == null) {
+			if (rememberMeServices instanceof AbstractRememberMeServices) {
+			    key = ((AbstractRememberMeServices) rememberMeServices).getKey();
```
Comment (eleftherias) [inline config/src/main/java/org/springframework/security/config/annotation/web/configurers/RememberMeConfigurer.java]: Please use a tab for indentation instead of spaces.

Comment (XieEDeHeiShou) [reply config/src/main/java/org/springframework/security/config/annotation/web/configurers/RememberMeConfigurer.java]: Done

**Comment by eleftherias** [comment]
Thank you for the changes @XieEDeHeiShou. Please add a test for any changes in this PR.
Let me know if you need help with writing the test.

**Comment by XieEDeHeiShou** [comment]
@eleftherias I'd like to provide test cases for these changes, and I do need some help, can you please guide me?

File: config/src/main/java/org/springframework/security/config/annotation/web/configurers/RememberMeConfigurer.java [hunk 4]
```diff
@@ -267,6 +267,11 @@ public void init(H http) throws Exception {
 		validateInput();
 		String key = getKey();
 		RememberMeServices rememberMeServices = getRememberMeServices(http, key);
+		if (key == null) {
```
[the remark this answers is not in the document]
Comment (XieEDeHeiShou) [reply config/src/main/java/org/springframework/security/config/annotation/web/configurers/RememberMeConfigurer.java]: I noticed this problem yesterday, and I decide to introduce a new  boolean member to indicate if the key is provided by the user or its generated. Are you agree with this idea?

[the remark this answers is not in the document]
Comment (XieEDeHeiShou) [reply config/src/main/java/org/springframework/security/config/annotation/web/configurers/RememberMeConfigurer.java]: Done

File: config/src/main/java/org/springframework/security/config/annotation/web/configurers/RememberMeConfigurer.java [hunk 5]
```diff
@@ -434,6 +434,9 @@ private UserDetailsService getUserDetailsService(H http) {
 	 * @return the remember me key to use
 	 */
 	private String getKey() {
+		if (this.key == null && this.rememberMeServices instanceof AbstractRememberMeServices) {
```
Comment (eleftherias) [inline config/src/main/java/org/springframework/security/config/annotation/web/configurers/RememberMeConfigurer.java]: Please move this check into the `null` check bellow, since we are already checking if `this.key == null`.

File: config/src/test/java/org/springframework/security/config/annotation/web/configurers/RememberMeConfigurerTests.java [hunk 6]
```diff
@@ -453,4 +456,35 @@ public void configureGlobal(AuthenticationManagerBuilder auth) throws Exception
...
+						assertThat(auth).isInstanceOf(RememberMeAuthenticationToken.class)));
+	}
+
+	@EnableWebSecurity
+	static class FallbackRememberMeKeyConfig extends RememberMeConfig {
+
+		@Override
+		protected void configure(HttpSecurity http) throws Exception {
+			// @formatter:off
+			http.rememberMe()
```
Comment (eleftherias) [inline config/src/test/java/org/springframework/security/config/annotation/web/configurers/RememberMeConfigurerTests.java]: I believe I misled you in my previous review, we will need additional configuration here.
```
// @formatter:off
http
	.authorizeRequests()
		.anyRequest().hasRole("USER")
		.and()
	.formLogin()
		.and()
	.rememberMe()
		.rememberMeServices(new TokenBasedRememberMeServices("key", userDetailsService()));
// @formatter:on
```

File: config/src/test/java/org/springframework/security/config/annotation/web/configurers/RememberMeConfigurerTests.java [hunk 7]
```diff
@@ -189,7 +191,8 @@ protected void configure(HttpSecurity http) throws Exception {
 		@Bean
 		public UserDetailsService userDetailsService() {
 			return new InMemoryUserDetailsManager(
-					User.withDefaultPasswordEncoder()
```
Comment (eleftherias) [inline config/src/test/java/org/springframework/security/config/annotation/web/configurers/RememberMeConfigurerTests.java]: Is this change necessary? `User.withDefaultPasswordEncoder` is deprecated, but it is acceptable to use in tests.

Comment (XieEDeHeiShou) [reply config/src/test/java/org/springframework/security/config/annotation/web/configurers/RememberMeConfigurerTests.java]: IDEA told me there is a warning when I commit this file, so I changed it. Should I revert this change?

Comment (eleftherias) [reply config/src/test/java/org/springframework/security/config/annotation/web/configurers/RememberMeConfigurerTests.java]: Yes, please revert this change. 
The IDE is correct to show a warning, but in this specific case it is allowed.

Comment (XieEDeHeiShou) [reply config/src/test/java/org/springframework/security/config/annotation/web/configurers/RememberMeConfigurerTests.java]: Done

**Comment by eleftherias** [comment]
Thanks for the PR @XieEDeHeiShou! This is now merged into master.
